Return MergeObserver from IndexWriter.forceMergeDeletes() (#14515)

salvatorecampagna opened the pull request:
## Summary

Changes `IndexWriter.forceMergeDeletes()` to return `MergePolicy.MergeObserver` instead of void, allowing applications to monitor merge progress and wait for completion without exposing mutable internal objects.

Fixes #14515

## Motivation

Currently when calling `forceMergeDeletes(false)`, there's no way to monitor whether merges were scheduled or wait for them to complete. This makes it difficult to coordinate merge completion with other operations or implement custom scheduling strategies.

## Changes

Added `MergePolicy.MergeObserver` as a new public API with methods to:
- Query merge count (`numMerges()`)
- Track completion progress (`numCompletedMerges()`)
- Wait synchronously with optional timeout (`await()`, `await(timeout, unit)`)
- Wait asynchronously via CompletableFuture (`awaitAsync()`)

The observer is thread-safe and handles cases where no merges are needed. It provides observability without exposing mutable `OneMerge` objects, maintaining proper encapsulation.

Updated `IndexWriter.forceMergeDeletes()` methods to return the observer instead of void. Also updated `RandomIndexWriter` to propagate the new return type.

## Testing

Added comprehensive tests covering:
- Blocking and non-blocking merge modes
- Timeout handling with proper assertions
- Completion status verification
- Null spec cases (no merges needed)
- Progress tracking via `numCompletedMerges()`

All tests pass with randomized seeds across multiple iterations.

## Backward Compatibility

Fully backward compatible. Existing code that ignores the return value continues to work unchanged. Changing void to a return type is not a breaking change in Java.

## Files Modified

- `MergePolicy.java` - New MergeObserver class with progress tracking
- `IndexWriter.java` - Updated return types and javadoc
- `RandomIndexWriter.java` - Propagated return type
- `TestIndexWriterMerging.java` - Comprehensive test coverage
- `CHANGES.txt` - Added entry under 10.4.0 API Changes

salvatorecampagna committed:
Return MergeObserver from IndexWriter.forceMergeDeletes() (#14515)

IndexWriter.forceMergeDeletes() now returns MergePolicy.MergeObserver
instead of void, allowing applications to monitor merge progress and
wait for completion. This enables coordination between merge completion
and other application logic, and supports both synchronous (await) and
asynchronous (CompletableFuture) waiting patterns.

Key capabilities:
- Query merge status: hasNewMerges(), numMerges()
- Wait synchronously: await(), await(timeout, unit)
- Wait asynchronously: awaitAsync() returns CompletableFuture<Void>
- Inspect individual merges: getMerge(int)

Changes:
- Add MergePolicy.MergeObserver nested class
- Update IndexWriter.forceMergeDeletes() methods to return MergeObserver
- Update RandomIndexWriter to propagate return type
- Add comprehensive tests (blocking/non-blocking modes, timeout handling)

Backward compatible: existing code that ignores the return value
continues to work without modification.

salvatorecampagna committed:
Add timing assertions to detect stuck merges

Measure actual merge time in testForceMergeDeletesWithObserver and
testMergeObserverAwaitWithTimeout, failing if merges take longer than
30 seconds. This helps detect stuck merges or performance regressions
while being generous enough for slow CI machines.

Both tests maintain their upper timeouts (implicit for await() and
10 minutes for await(timeout, unit)) to prevent indefinite hangs,
but now also fail fast at 30 seconds if something is wrong.

vigyasharma commented on lucene/core/src/java/org/apache/lucene/index/MergePolicy.java:
```diff
@@ -939,4 +939,96 @@ static final class MergeReader {
+     * @return the merge at index i
+     * @throws IndexOutOfBoundsException if index is invalid or no merges exist
+     */
+    public MergePolicy.OneMerge getMerge(int i) {
```
I'm not sure if providing this method is really useful. The caller can't know the "index" location of a OneMerge unless they already have access to the MergeSpec.

vigyasharma replied:
Also, I don't think that the `OneMerge` object is immutable?

vigyasharma commented on lucene/core/src/test/org/apache/lucene/index/TestIndexWriterMerging.java:
```diff
@@ -333,6 +337,330 @@ public synchronized void merge(MergeSource mergeSource, MergeTrigger trigger)
+
+    assertTrue(
+        "Merge took too long: " + elapsedMillis + "ms (expected < 30000ms)",
+        elapsedMillis < 30_000);
```
Is this 30_000 a random magic number for timeout upperbound? Why don't we just use it in the await timeout value? Then `assertTrue` will fail if we don't meet this limit.

vigyasharma commented on lucene/core/src/test/org/apache/lucene/index/TestIndexWriterMerging.java:
```diff
@@ -333,6 +337,330 @@ public synchronized void merge(MergeSource mergeSource, MergeTrigger trigger)
+
+    MergePolicy.MergeObserver observer = iw.forceMergeDeletes(true);
+    assertTrue("Should have scheduled merges", observer.hasNewMerges());
+    assertTrue("Should have completed merges", observer.numMerges() > 0);
```
You can add an assert here that tests that numMerges == completedMerges

vigyasharma commented on lucene/CHANGES.txt:
```diff
@@ -69,6 +69,12 @@ Improvements
 
 * GITHUB#15124: Use RamUsageEstimator to calculate size for non-accountable queries. (Sagar Upadhyaya)
 
+* GITHUB#14515: IndexWriter.forceMergeDeletes() now returns MergePolicy.MergeObserver,
```
This change can go into 10.4. Let's move this entry to the "API Changes" section under 10.4?

vigyasharma replied:
We add the changes entry in the Lucene version we are targeting. All changes go to `main` branch regardless, but the ones for a specific target branch (like 10.4 in this case) additionally get backported to 10.x.

So you should just add the entry to 10.4 in this PR itself and remove it from 11.0. After I merge this into main, i'll backport the commit to 10.x branch. Will reach out to you if I need any help with backport.

vigyasharma commented on lucene/core/src/test/org/apache/lucene/index/TestIndexWriterMerging.java:
```diff
@@ -333,6 +337,330 @@ public synchronized void merge(MergeSource mergeSource, MergeTrigger trigger)
     public void close() {}
   }
 
```
Appreciate the thorough tests below, nice work!

vigyasharma reviewed (commented):
I like this approach, it let's us observe merge progress without exposing mutable objects. Thanks for making these changes @salvatorecampagna

salvatorecampagna committed:
Remove getMerge() method from MergeObserver

The getMerge(int) method exposed mutable OneMerge objects, breaking
encapsulation. Callers also cannot meaningfully use an index-based API
without already having the MergeSpecification.

Refactored IndexWriter to work with MergeSpecification directly for
merge registration and waiting. Removed associated tests.

salvatorecampagna committed:
Remove hasNewMerges() method from MergeObserver

The hasNewMerges() method is confusing - it's unclear whether it
counts in-progress merges, completed merges, etc. Users can simply
check numMerges() > 0 instead for better clarity.

This change:
- Removes the hasNewMerges() method from MergeObserver
- Updates all test cases to use numMerges() > 0 instead
- Removes redundant assertions where both hasNewMerges() and
  numMerges() were checked

salvatorecampagna committed:
Add numCompletedMerges() method to MergeObserver

This method allows tracking merge progress by checking how many merges
have completed out of the total. Users can calculate progress as:
  int progress = (int)(100.0 * observer.numCompletedMerges() / observer.numMerges());

This is particularly useful when monitoring long-running merge operations
such as forceMerge or forceMergeDeletes with non-blocking mode.

Also updates toString() to display the completion count for better
debugging visibility.

salvatorecampagna committed:
Fix timeout assertions and add completion checks in tests

Instead of using magic numbers for timeout assertions, use the actual
timeout value in the await() call. This makes the tests clearer and
ensures the timeout is enforced properly.

Also adds assertions to verify that all merges are completed after
await() returns true, leveraging the new numCompletedMerges() API.

Changes:
- testForceMergeDeletesWithObserver: Use 30s timeout in await() call
- testMergeObserverAwaitWithTimeout: Use 30s timeout in await() call
- Both tests: Add assertion that numMerges == numCompletedMerges

salvatorecampagna committed:
Move CHANGES.txt entry to 10.4.0 API Changes section

The forceMergeDeletes() API change should be documented in the 10.4.0
release (the target version for this feature) under API Changes, not
in the 11.0.0 Improvements section.

Moved the entry from:
- 11.0.0 → Improvements
To:
- 10.4.0 → API Changes

salvatorecampagna commented on lucene/core/src/java/org/apache/lucene/index/MergePolicy.java:
```diff
@@ -939,4 +939,96 @@ static final class MergeReader {
+     *
+     * @return number of merges, or 0 if no merges were scheduled
+     */
+    public int numMerges() {
```
Excellent suggestion! This is a far better way to provide progress tracking without exposing mutable `OneMerge` objects. I'll add `numCompletedMerges()` that iterates through the merges and checks mergeCompleted.isDone().

Users can then track progress as: `100 * numCompletedMerges() / numMerges()`

I'll also update `toString()` to show the completion count for better debugging.

salvatorecampagna committed:
Remove completion count from MergeObserver.toString()

The toString() method was calling numCompletedMerges(), which iterates
through all merges to count completed ones. This is problematic because
toString() is frequently called by debuggers, logging frameworks, and IDE
watches, making the O(n) iteration too expensive for a debugging method.

Users who need the completion count can explicitly call the
numCompletedMerges() method. The toString() should remain minimal and
cheap for debugging purposes.

salvatorecampagna committed:
Minimize IndexWriter changes to preserve original code structure

The previous refactoring in commit 05b82666dc8 changed more than
necessary in IndexWriter, including switching from indexed for-loops
to enhanced for-loops and removing the newMergesFound variable. This
commit restores the original code structure while still avoiding the
removed getMerge() method.

Changes:
- Restore newMergesFound boolean variable as in original code
- Use indexed for-loops as in original code
- Access spec.merges directly instead of removed getMerge()
- Only add the return statement needed for the new API

This keeps the diff minimal and preserves the original code patterns,
making the change easier to review and maintain.

salvatorecampagna committed:
Apply code formatting from gradle tidy

Reformats assertTrue statements to single line as per project style.

salvatorecampagna committed:
Fix forbidden String.format() usage in MergeObserver.toString()

Use String.format() with Locale.ROOT to avoid locale-dependent formatting.
This is the standard pattern used throughout Lucene codebase.

vigyasharma commented on lucene/core/src/java/org/apache/lucene/index/MergePolicy.java:
```diff
@@ -939,4 +940,93 @@ static final class MergeReader {
+      if (spec == null) {
+        return "MergeObserver: no merges";
+      }
+      return String.format(Locale.ROOT, "MergeObserver: %d merges", numMerges());
```
We could include `spec.toString()` here. It has some useful info about the segments in each merge.

salvatorecampagna replied:
According to a test I executed locally the outout should be something like. Are we sure we want this level of detail? I guess the output might be long if there are meny merges.

```
MergeObserver: 3 merges
MergeSpec:
  1: _0(11.0.0):c5 _1(11.0.0):c5 into _k
  2: _2(11.0.0):c5 _3(11.0.0):c5 into _l
  3: _4(11.0.0):c5 _5(11.0.0):c5 into _m
```

salvatorecampagna replied:
My concern is that `spec.toString()` was internal before, so we controlled where it's used. Now that `MergeObserver` is public API, users might call `toString()` frequently (logging, monitoring) without realizing it could be expensive with many merges.

salvatorecampagna committed:
Include segment details in MergeObserver.toString()

Update MergeObserver.toString() to include the underlying MergeSpec
details by appending spec.toString() to the merge count. This provides
useful information about segments in each merge.

Sample output with 3 merges:
```
MergeObserver: 3 merges
MergeSpec:
  1: _0(11.0.0):c5 _1(11.0.0):c5 into _k
  2: _2(11.0.0):c5 _3(11.0.0):c5 into _l
  3: _4(11.0.0):c5 _5(11.0.0):c5 into _m
```

salvatorecampagna committed:
Apply code formatting from gradle tidy

Splits long String.format() call across multiple lines for readability
as per project style guidelines.

vigyasharma reviewed (approved):
Looks great, thanks @salvatorecampagna . And congratulations on your first contribution!
